crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -1,373 +1,773 @@
-                another parameter with mutable access to the same `{res}` resource.",
-                system_state.name,
-                res = std::any::type_name::<T>()
-            );
+                "NonSend<{}> in system {} conflicts with a previous mutable resource access ({0}). Allowing this would break Rust's mutability rules. Consider removing the duplicate access.",
+                std::any::type_name::<T>(), system_state.name);
+        }
+        combined_access.add_read(component_id);
+
+        let resource_archetype = world.archetypes.resource();
+        let archetype_component_id = resource_archetype
+            .get_archetype_component_id(component_id)
+            .unwrap();
+        system_state
+            .archetype_component_access
+            .add_read(archetype_component_id);
+        Self {
+            component_id,
+            marker: PhantomData,
         }
-        system_state.resource_access.add_read(TypeId::of::<T>());
     }
+}
+
+impl<'a, T: 'static> SystemParamFetch<'a> for NonSendState<T> {
+    type Item = NonSend<'a, T>;
 
     #[inline]
     unsafe fn get_param(
+        state: &'a mut Self,
         _system_state: &'a SystemState,
-        _world: &'a World,
-        resources: &'a Resources,
+        world: &'a World,
     ) -> Option<Self::Item> {
-        let (value, added, mutated) =
-            resources.get_unsafe_ref_with_added_and_mutated::<T>(ResourceIndex::Global);
-        if *added.as_ptr() || *mutated.as_ptr() {
-            Some(ChangedRes::new(value))
-        } else {
-            None
-        }
+        Some(NonSend {
+            value: world.get_non_send_with_id::<T>(state.component_id)?,
+        })
     }
 }
 
-pub struct FetchLocal<T>(PhantomData<T>);
+/// Unique borrow of a NonSend resource
+pub struct NonSendMut<'a, T: 'static> {
+    pub(crate) value: &'a mut T,
+    pub(crate) flags: &'a mut ComponentFlags,
+}
+
+impl<'a, T: 'static> Deref for NonSendMut<'a, T> {
+    type Target = T;
 
-impl<'a, T: Resource + FromResources> SystemParam for Local<'a, T> {
-    type Fetch = FetchLocal<T>;
+    #[inline]
+    fn deref(&self) -> &T {
+        self.value
+    }
 }
-impl<'a, T: Resource + FromResources> FetchSystemParam<'a> for FetchLocal<T> {
-    type Item = Local<'a, T>;
 
-    fn init(system_state: &mut SystemState, _world: &World, resources: &mut Resources) {
-        if system_state
-            .local_resource_access
-            .is_read_or_write(&TypeId::of::<T>())
-        {
+impl<'a, T: 'static> DerefMut for NonSendMut<'a, T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut T {
+        self.flags.insert(ComponentFlags::MUTATED);
+        self.value
+    }
+}
+
+impl<'a, T: 'static + core::fmt::Debug> core::fmt::Debug for NonSendMut<'a, T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        self.value.fmt(f)
+    }
+}
+
+pub struct NonSendMutFetch<T>(PhantomData<T>);
```

Comment:
This struct is unused

Reply:
Not just that, we don't need to differentiate `NonSend` and `NonSendMut`: only one thread at a time can ever access something that is `!Send`, so we can give it exclusive (mutable) access off the bat.

Reply:
You can give multiple shared references to something at the same time, so `NonSend` can happen multiple times for the same thing, but only one exclusive reference can exist at the same time, so only one `NonSendMut` for the same thing can be used at a time.

Reply:
How could there be multiple shared references? A thread can run only one system at a time, systems are always ran to completion, and a resource inserted as `!Send` can only be accessed from the main thread. (Although, I haven't seen yet if there were any changes to the implementation.)

Reply:
If a single system uses the same `NonSend` system param twice, that would work. If it used the same `NonSendMut` system param twice, it would panic to prevent two mutable borrows of the same thing.

Reply:
Providing only exclusive access implies we check for such aliasing, of course. Is there a legitimate case for a single system to ask for multiple shared references to the same non-send resource?

Reply:
A user may use multiple system param "bundles" that reference among other things the same non-send resource.

Reply:
That's... a bit contrived; but, yes, we shouldn't make this impossible just because. For the record, I would've been content with "API parity is good" 😄

Reply:
removed!